pkg/daemon/controller/remote.go

```diff
@@ -0,0 +1,177 @@
+package controller
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+
+	ramav1 "github.com/oecp/rama/pkg/apis/networking/v1"
+	"github.com/oecp/rama/pkg/daemon/containernetwork"
+	"github.com/vishvananda/netlink"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog"
+)
+
+type RemoteVtepCache struct {
+	remoteVtepIPMap map[string]net.HardwareAddr
+
+	ch chan struct{}
+}
+
+func NewRemoteVtepCache() *RemoteVtepCache {
+	return &RemoteVtepCache{
+		remoteVtepIPMap: map[string]net.HardwareAddr{},
+		ch:              make(chan struct{}, 1),
+	}
+}
+
+func (rvc *RemoteVtepCache) UpdateRemoteVtepIPs(remoteVteps []*ramav1.RemoteVtep) error {
+	rvc.lock()
+	defer rvc.unlock()
+
+	rvc.remoteVtepIPMap = map[string]net.HardwareAddr{}
+
+	for _, remoteVtep := range remoteVteps {
+		macAddr, err := net.ParseMAC(remoteVtep.Spec.VtepMAC)
+		if err != nil {
+			return fmt.Errorf("parse node vtep mac %v failed: %v", remoteVtep.Spec.VtepMAC, err)
+		}
+
+		rvc.remoteVtepIPMap[remoteVtep.Spec.VtepIP] = macAddr
+	}
+
+	return nil
+}
+
+func (rvc *RemoteVtepCache) SearchIP(ip net.IP) (net.HardwareAddr, bool) {
+	rvc.lock()
+	defer rvc.unlock()
+
+	mac, exist := rvc.remoteVtepIPMap[ip.String()]
+	return mac, exist
+}
+
+func (rvc *RemoteVtepCache) lock() {
+	rvc.ch <- struct{}{}
+}
+
+func (rvc *RemoteVtepCache) unlock() {
+	<-rvc.ch
+}
+
+// add handler for RemoteVtep and RemoteSubnet
+
+func (c *Controller) enqueueAddOrDeleteRemoteVtep(obj interface{}) {
+	c.nodeQueue.Add(ActionReconcileNode)
+}
+
+func (c *Controller) enqueueUpdateRemoteVtep(oldObj, newObj interface{}) {
+	oldRv := oldObj.(*ramav1.RemoteVtep)
+	newRv := newObj.(*ramav1.RemoteVtep)
+
+	if oldRv.Spec.VtepIP != newRv.Spec.VtepIP ||
+		oldRv.Spec.VtepMAC != newRv.Spec.VtepMAC || !reflect.DeepEqual(oldRv.Status.PodIPList, newRv.Status.PodIPList) {
```

Comment:
Slice is unsorted. If using deepequal, two slice having same value maybe false. Using set instead. For example, "github.com/gogf/gf/container/gset"

Reply:
Nice I've thought about that before. Is reordering of pod ip list a frequent op? If not, I believe deepEqual is tolerable.

Reply:
Reordering cannot be guaranteed. Besides, reflect is slow. Try the set and use diff fucntion, it's efficient.

Reply:
OK, I shall follow that.